### PR TITLE
Upgrade rust toolchain to nightly-2025-01-01

### PR DESCRIPTION
**Pull request by Rahix**
Time for another toolchain upgrade to fix a number of known compiler bugs.

### CODE REVIEW

**Comment by Rahix** [comment]
@Patryk27, some CI jobs are failing with

```
out of range branch target (expected an integer in the range -4096 to 4095)
```

is this still one of the known problems or something new?

**Comment by Patryk27** [comment]
This is known, it’s already fixed in LLVM, but not yet backported - it should be in rustc within a couple days hopefully; I’ll let you know, so we can wait before bumping the toolchain here to get all of the fixes 🙂

**Comment by Patryk27** [comment]
@Rahix I think we should be good now - could you bump to the newest toolchain? :eyes:

**Comment by Rahix** [comment]
@Patryk27, unfortunately it looks like we are still hitting the dreaded `error: out of range branch target` with `nightly-2024-10-11`? :disappointed:

Just to check, I also tried `nightly-2024-10-12`, but I can still reproduce the error message...

**Comment by Patryk27** [comment]
Ouch, my fault! It seems that https://github.com/llvm/llvm-project/pull/109124 wasn't backported - I'll handle it tomorrow. Other commits seem to be in place, though (e.g. https://github.com/Rahix/avr-hal/issues/573 is fixed).

**Comment by Rahix** [comment]
No worries!  Just ping me once the patch has hit nightly Rust, then I'll give it another shot :)

Thanks for all your compiler work!

**Comment by Patryk27** [comment]
Alright, let's try now - the appropriate fix should've gone with https://github.com/rust-lang/rust/pull/132352.

**Comment by Rahix** [comment]
Updated to `nightly-2024-11-03`, but CI still reports the `error: out of range branch target` for some targets.  Disabled `fail-fast` so we get the full list of failing targets - I checked, they are all failing on the same error.

Is the fix maybe not actually in this version I tried?

**Comment by Patryk27** [comment]
Could you try `2024-11-05`? 
The fix is there, but if it got merged on `2024-11-02`, so maybe the `2024-11-03` build doesn't contain it just yet 👀 

Either that or it's going to be a new kind of error, we'll see 😄

**Comment by Rahix** [comment]
Updated to 2024-11-05, still same targets failing... So I guess this is a new thing then? :sweat_smile:

**Comment by Patryk27** [comment]
Looks this way 😅 I'll try reproducing locally and report back.

**Comment by Patryk27** [comment]
Status: reproduced!

```
$ cd ~/avr-hal/mcu/attiny-hal
$ RUSTFLAGS='--emit=llvm-ir' cargo +stage1 build --features attiny85 -Z build-std=core --target ../../avr-specs/avr-attiny85.json --release
$ llc -march=avr -mcpu=attiny85 -filetype=obj ~/avr-hal/target/avr-attiny85/release/deps/core-*.ll
```

... will crash, saying `out of range branch target`.

I've altered the error message in LLVM to print the offset it's trying to generate instruction for and it's pretty obvious something's off:

```
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4488)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4336)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4220)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4198)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 102544)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 101922)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 101092)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 99796)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 99334)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6342)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6300)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6246)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6180)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6146)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6148)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5200)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5030)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4876)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4556)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4352)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4278)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5154)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5112)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5078)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5046)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 5022)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4210)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 9260)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 9190)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 9130)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 9074)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 9010)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 9012)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 8940)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 7492)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 7050)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6696)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 6792)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4472)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4440)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 4434)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 102228)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 102180)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709547180)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709546970)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709546294)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709547508)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709547222)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709546918)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709546776)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709546334)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709544838)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709546100)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709544604)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709545552)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709544062)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18066)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 16516)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 15446)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 14552)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 12980)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 11236)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 10048)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 8678)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709504680)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709503924)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709496220)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709492994)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709429400)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709429294)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709429132)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709429028)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 19688)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 7794)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709481512)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709484008)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709488226)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709481262)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709483846)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709484658)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709487990)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709484552)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709487880)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709479190)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709482250)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709488688)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709478826)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709482044)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709485504)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709488394)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709485370)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709488258)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709486348)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709483578)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709475852)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709544056)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709543836)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709543560)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709543340)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709534116)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709537724)
<unknown>:0: error: out of range branch target (expected an integer in the range -4096 to 4095, got 18446744073709429922)
```

I'll try narrowing the issue down 🫡

**Comment by Patryk27** [comment]
Got it!

Reduced:
https://gist.github.com/Patryk27/a68ed01eeb57b59c3025c395059f906d
(and then `llc -march=avr -mcpu=attiny85 -filetype=obj core.reduced.ll`)

The issue boils down to too large jumps - basically, given:

```asm
foo:
  rjmp bar

# lots of code

bar:
  # something
```

... if there's too many opcodes between `rjmp` and the target label, we might not have enough space in the opcode's encoding to represent the offset (`rjmp` allows for 12 bits, for instance).

To avoid this issue, LLVM has a pass called _branch relaxation_ that detects too large jumps and upgrades them into different instructions - in the case of AVR, that's where the backend might decide to emit `jmp` instead; it's one clock cycle more, but allows for a larger offset.

Everything's fine if the target AVR supports `jmp` - attiny85 does not, so the backend decides to stick with `rjmp`:

https://github.com/llvm/llvm-project/blob/a76609dd72743c0d678504042b00d434e6cebe1a/llvm/lib/Target/AVR/AVRInstrInfo.cpp#L568

... just to crash a moment later, in sort of a ¯\_(ツ)_/¯ whaddya-want-me-to-do fashion.

Fortunately, not all hope is lost! - there's two ways out of this situation:

- we can exploit the fact that AVR memory wraps, so instead of generating e.g. `rjmp +5000`, we can emit `rjmp -3192` (cursed, but practical),
- we can emit `ijmp` (this would be a rather heavy approach, though, since we'd have to store the `Z` register somewhere, update it, do the jump and then restore the register at the target site).

We didn't have to worry about those offsets, because in the past - up until https://github.com/llvm/llvm-project/commit/6859685a87ad093d60c8bed60b116143c0a684c7 aka https://github.com/llvm/llvm-project/pull/106722 - LLVM handled jumps by relying on relocations. Relocations allow to store larger offsets with the assumption that "the linker will somehow deal with them later", and it did: (most likely) by applying the wrapping trick.

tl;dr linker smart, linker can solve "large jumps" issue - but llvm no longer rely on linker for jump, llvm must be taught to emit long jumps on its own

I'll try preparing the patch in the upcoming days.

**Comment by Rahix** [comment]
Damn, nice work!

> we can exploit the fact that AVR memory wraps, so instead of generating e.g. rjmp +5000, we can emit rjmp -3192 (cursed, but practical)

haha, i love it :laughing: cursed, indeed

**Comment by Rahix** [comment]
@Patryk27, thanks again for your work fixing the latest iteration of the compiler error.  I assume you're tracking when https://github.com/llvm/llvm-project/pull/118015 will hit rustc nightly?  Just need to know when I can move this PR here forward once again :)

**Comment by Patryk27** [comment]
Yes, I'm tracking it 🙂 

fwiw, I think we'll need to wait for https://github.com/llvm/llvm-project/pull/121498 as well (caught over https://github.com/llvm/llvm-project/pull/118015#issuecomment-2564055124).

**Comment by tones111** [comment]
The regression fixes we've been waiting on have landed in LLVM 20 which is currently being integrated into rustc in rust-lang/rust#135763.

**Comment by jaudiger** [comment]
The MR has been merged an hour ago. It should be available soon in nightlies builds. This MR is almost unblocked !

**Comment by Rahix** [comment]
Superseded by #634.
